mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -1647,13 +1661,61 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
         padGemmFilterTransform.passThrough("gemmN");
       }
 
-      TransformMapAttr padGemmFilterTransformAttr =
-          padGemmFilterTransform.get();
+      padGemmFilterTransformAttr = padGemmFilterTransform.get();
       gemmFilterPad = b.create<miopen::TransformOp>(loc, gemmFilter,
                                                     padGemmFilterTransformAttr);
       // filter pad end
     }
 
+    // KPack for filter tensor.
+    Value gemmFilterKPack = gemmFilterPad;
+
+    // FIXME. consider backward convolution.
+    if ((KPack > 1) && (convOpType == miopen::ConvOpType::Fwd)) {
+      if (!isFilterPad) {
+        BottomUpCTBuilder kpackGemmFilterTransform =
+            BottomUpCTBuilder::above(filterTransform, filterTransformAttr);
+        int64_t gemmKLength = filterTransform.endSize("gemmK");
+        kpackGemmFilterTransform.passThrough("gemmG");
+        kpackGemmFilterTransform.passThrough("gemmM");
+        kpackGemmFilterTransform.unmerge({"gemmK", "gemmKPack"}, {1, 3},
+                                         "gemmK", {gemmKLength / KPack, KPack});
+        TransformMapAttr kpackGemmFilterTransformAttr =
+            kpackGemmFilterTransform.get();
+        gemmFilterKPack = b.create<miopen::TransformOp>(
+            loc, gemmFilter, kpackGemmFilterTransformAttr);
+      } else {
+        // isFilterPad == true.
+        BottomUpCTBuilder kpackGemmFilterTransform = BottomUpCTBuilder::above(
+            padGemmFilterTransform, padGemmFilterTransformAttr);
+        int64_t gemmKLength = 0;
+        if (filterCheckPadGemmK) {
+          gemmKLength = padGemmFilterTransform.endSize("gemmKPad");
+        } else {
+          gemmKLength = padGemmFilterTransform.endSize("gemmK");
+        }
+        kpackGemmFilterTransform.passThrough("gemmG");
+        if (filterCheckPadGemmM) {
+          kpackGemmFilterTransform.passThrough("gemmMPad");
+        } else {
+          kpackGemmFilterTransform.passThrough("gemmM");
+        }
+        if (filterCheckPadGemmK) {
+          kpackGemmFilterTransform.unmerge({"gemmKPad", "gemmKPack"}, {1, 3},
+                                           "gemmKPad",
+                                           {gemmKLength / KPack, KPack});
+        } else {
+          kpackGemmFilterTransform.unmerge({"gemmK", "gemmKPack"}, {1, 3},
+                                           "gemmK",
+                                           {gemmKLength / KPack, KPack});
+        }
+        TransformMapAttr kpackGemmFilterTransformAttr =
+            kpackGemmFilterTransform.get();
+        gemmFilterKPack = b.create<miopen::TransformOp>(
+            loc, gemmFilterPad, kpackGemmFilterTransformAttr);
+      }
+    }
+
```

Comment:
I think there's a much simpler way to do this, using `endName` on `padGemmFilterTransform` (which might just be `gemmFilterTransform`). That way, you can grab the relevant dimension names, whether they're "gemmXPad" or "gemmX" and don't have to have this pile of branches that doesn't add much.